pallets/pips/src/lib.rs

```diff
@@ -1135,148 +1249,113 @@ impl<T: Trait> Module<T> {
-            if proposal.state == ProposalState::Referendum {
-                match Self::check_beneficiaries(id) {
-                    Ok(_) => {
-                        match proposal.proposal.dispatch(system::RawOrigin::Root.into()) {
-                            Ok(post_info) => {
-                                actual_weight = post_info.actual_weight.unwrap_or(0);
-                                Self::pay_to_beneficiaries(id);
-                                Self::update_referendum_state(id, ReferendumState::Executed);
-                            }
-                            Err(e) => {
-                                Self::update_referendum_state(id, ReferendumState::Failed);
-                                debug::error!(
-                                    "Referendum {}, its execution fails: {:?}",
-                                    id,
-                                    e.error
-                                );
-                            }
-                        };
-                    }
-                    Err(e) => {
-                        Self::update_referendum_state(id, ReferendumState::Failed);
-                        debug::error!("Referendum {}, its beneficiaries fails: {:?}", id, e);
-                    }
-                }
-                Self::prune_data(id, Self::prune_historical_pips());
-            }
-        }
-        actual_weight
-    }
-
-    fn check_beneficiaries(id: PipId) -> DispatchResult {
-        if let Some(proposal) = Self::proposals(id) {
-            if let Some(beneficiaries) = proposal.beneficiaries {
-                let total_amount = beneficiaries
-                    .iter()
-                    .fold(0.into(), |acc, b| b.amount.saturating_add(acc));
-                ensure!(
-                    T::Treasury::balance() >= total_amount,
-                    Error::<T>::InsufficientTreasuryFunds
-                );
-            }
-        }
-        Ok(())
     }
 
-    fn pay_to_beneficiaries(id: PipId) {
-        if let Some(proposal) = Self::proposals(id) {
-            if let Some(beneficiaries) = proposal.beneficiaries {
-                let _ = beneficiaries.into_iter().fold(0.into(), |acc, b| {
-                    T::Treasury::disbursement(b.id, b.amount);
-                    b.amount.saturating_add(acc)
-                });
+    /// Execute the PIP given by `id`.
+    /// Panics if the PIP doesn't exist or isn't scheduled.
+    fn execute_proposal(id: PipId) -> Weight {
+        let proposal = Self::proposals(id).expect("PIP was scheduled but doesn't exist");
+        assert_eq!(proposal.state, ProposalState::Scheduled);
+        let (new_state, weight) = match proposal.proposal.dispatch(system::RawOrigin::Root.into()) {
+            Ok(post_info) => (ProposalState::Executed, post_info.actual_weight),
+            Err(e) => {
+                debug::error!("Proposal {}, its execution fails: {:?}", id, e.error);
+                (ProposalState::Failed, None)
```

Comment:
If a proposal is returning the `DispatchResultWithPostInfo` then in error it will also return the actual weight. So you can replace the `None` by `e.post_info.actual_weight`.